File: dawn_cpp/func_rb.cpp
#include <iostream>
#include <memory>

enum class Color { Red, Black };

template <typename T>
struct RBTree {
    Color color;
    std::unique_ptr<RBTree<T>> left, right;
    T data;

    RBTree(Color c, std::unique_ptr<RBTree<T>> l, T d, std::unique_ptr<RBTree<T>> r)
        : color(c), left(std::move(l)), data(std::move(d)), right(std::move(r)) {}
};

template <typename T>
std::unique_ptr<RBTree<T>> balance(Color color, std::unique_ptr<RBTree<T>> a, T x, std::unique_ptr<RBTree<T>> b) {
    if (color == Color::Black) {
        if (a && a->color == Color::Red && a->left && a->left->color == Color::Red) {
            return std::make_unique<RBTree<T>>(
                Color::Red,
                std::move(a->left->left),
                std::move(x),
                std::make_unique<RBTree<T>>(
                    Color::Black, std::move(a->right), std::move(a->data), std::move(b)));
        }
        if (a && a->color == Color::Red && a->right && a->right->color == Color::Red) {
            return std::make_unique<RBTree<T>>(
                Color::Red,
                std::make_unique<RBTree<T>>(
                    Color::Black, std::move(a->left), std::move(a->data), std::move(a->right->left)),
                std::move(x),
                std::make_unique<RBTree<T>>(
                    Color::Black, std::move(a->right->right), std::move(b->data), std::move(b->right)));
        }
        if (b && b->color == Color::Red && b->right && b->right->color == Color::Red) {
            return std::make_unique<RBTree<T>>(
                Color::Red,
                std::make_unique<RBTree<T>>(
                    Color::Black, std::move(a), std::move(x), std::move(b->left)),
                std::move(b->data),
                std::move(b->right));
        }
        if (b && b->color == Color::Red && b->left && b->left->color == Color::Red) {
            return std::make_unique<RBTree<T>>(
                Color::Red,
                std::make_unique<RBTree<T>>(
                    Color::Black, std::move(a), std::move(x), std::move(b->left->left)),
                std::move(b->left->data),
                std::make_unique<RBTree<T>>(
                    Color::Black, std::move(b->left->right), std::move(b->data), std::move(b->right)));
        }
    }
    return std::make_unique<RBTree<T>>(color, std::move(a), std::move(x), std::move(b));
}

template <typename T>
std::unique_ptr<RBTree<T>> insert(std::unique_ptr<RBTree<T>> t, T x) {
    if (!t) {
        return std::make_unique<RBTree<T>>(Color::Red, nullptr, std::move(x), nullptr);
    }
    if (x < t->data) {
        t->left = insert(std::move(t->left), std::move(x));
    } else if (x > t->data) {
        t->right = insert(std::move(t->right), std::move(x));
    } else {
        return t;
    }
    return balance(t->color, std::move(t->left), std::move(t->data), std::move(t->right));
}

File: dawn_go/func_rb.go
package main

import "fmt"

type Color bool

const (
	Red   Color = true
	Black Color = false
)

type RBTree struct {
	color Color
	left  *RBTree
	data  int
	right *RBTree
}

func balance(color Color, a, b *RBTree, x int) *RBTree {
	if color == Black {
		if isRed(a) && isRed(a.left) {
			return &RBTree{Red, a.left.copy(), x, &RBTree{Black, a.right.copy(), a.data, b.copy()}}
		}
		if isRed(a) && isRed(a.right) {
			return &RBTree{Red, &RBTree{Black, a.left.copy(), a.data, a.right.left.copy()}, x, &RBTree{Black, a.right.right.copy(), b.data, b.right.copy()}}
		}
		if isRed(b) && isRed(b.right) {
			return &RBTree{Red, &RBTree{Black, a.copy(), x, b.left.copy()}, b.data, b.right.copy()}
		}
		if isRed(b) && isRed(b.left) {
			return &RBTree{Red, &RBTree{Black, a.copy(), x, b.left.left.copy()}, b.left.data, &RBTree{Black, b.left.right.copy(), b.data, b.right.copy()}}
		}
	}
	return &RBTree{color, a.copy(), x, b.copy()
}

func isRed(t *RBTree) bool {
	if t == nil {
		return false
	}
	return t.color == Red
}

func (t *RBTree) copy() *RBTree {
	if t == nil {
		return nil
	}
	return &RBTree{t.color, t.left.copy(), t.data, t.right.copy()}
}

func insert(t *RBTree, x int) *RBTree {
	if t == nil {
		return &RBTree{Red, nil, x, nil}
	}
	if x < t.data {
		t.left = insert(t.left, x)
	} else if x > t.data {
		t.right = insert(t.right, x)
	} else {
		return t
	}
	return balance(t.color, t.left, t.right, t.data)
}